### PR TITLE
[CALCITE-3864] Supports CONCAT for variable arguments

### PR DESCRIPTION
**Pull request by wenhuitang**
Pull request for https://issues.apache.org/jira/browse/CALCITE-3864

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -460,6 +460,15 @@ public static ByteString concat(ByteString s0, ByteString s1) {
     return s0.concat(s1);
   }
 
+  /** SQL function CONCAT(arg, ...)*/
```
Comment (XuQianJin-Stars) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: I noticed that the number of function parameters in different databases is different:
Mysql 
```
CONCAT(string1,string2, ... );
```
Oracle
```
CONCAT(char1, char2)
```
PostgreSql
```
CONCAT(str_1, str_2, ...)
```

Comment (wenhuitang) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: I noticed this problem too, and I found that creating two SqlFunction for CONCAT is redundant and unnecessary, what's more, it maybe cause exception when look up method.
I have updated this pull request, concat function is equivalent to || operator.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 2]
```diff
@@ -2169,6 +2170,21 @@ protected static Calendar getCalendarNotTooNear(int timeUnit) {
         "VARCHAR(33335) NOT NULL");
     tester.checkNull("x'ff' || cast(null as varbinary)");
     tester.checkNull(" cast(null as ANY) || cast(null as ANY) ");
+    tester.checkString("cast('a' as varchar) || cast('b' as varchar) "
+        + "|| cast('c' as varchar)", "abc", "VARCHAR NOT NULL");
+
+    final SqlTester tester1 = tester(SqlLibrary.MYSQL);
+    final SqlTester tester2 = tester(SqlLibrary.POSTGRESQL);
+    final SqlTester tester3 = tester(SqlLibrary.ORACLE);
```
Comment (XuQianJin-Stars) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: oracle `CONCAT (char1, char2)` only accept how to reflect two parameters?

Comment (wenhuitang) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: I don't think that it is a serious problem since concat function will be translated to || operator and it can work well for oracle. In a practical application scenario, we may use two SqlLibaray at a same time, if they have different concat function, ambiguity will be caused when check arguements, even we register oracle concat as a alias of SqlStdOperatorTable.CONCAT.

Comment (XuQianJin-Stars) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: > I don't think that it is a serious problem since concat function will be translated to || operator and it can work well for oracle and also support concat('a', 'b'). In a practical application scenario, we may use two SqlLibaray at a same time, if they have different concat function, ambiguity will be caused when check arguements, even we register oracle concat as a alias of SqlStdOperatorTable.CONCAT.

I did not say that this is a serious problem, but I noticed that oracle's `concat` is a special case of your `concat` implementation. Because oracle `concat` only supports two parameters.

**Comment by XuQianJin-Stars** [comment]
hi @wenhuitang In addition, you need to add function docs.

**Comment by wenhuitang** [comment]
> hi @wenhuitang In addition, you need to add function docs.

the description of CONCAT(string [, string ]*)  has already existed in Line 2325 of the document reference.md

File: core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java [hunk 3]
```diff
@@ -208,12 +212,52 @@ public boolean supportsAliasedValues() {
 
       unparseFloor(writer, call);
       break;
-
     default:
+      if (call.getOperator() == SqlStdOperatorTable.CONCAT) {
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java]: Please move this piece of code before `switch-case` block, you can refer to the style in `OracleSqlDialect`.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 4]
```diff
@@ -286,8 +286,13 @@ private SqlLibraryOperators() {
               opBinding -> {
                 int precision = opBinding.collectOperandTypes().stream()
                     .mapToInt(RelDataType::getPrecision).sum();
-                return opBinding.getTypeFactory()
-                    .createSqlType(SqlTypeName.VARCHAR, precision);
+                if (precision > 0) {
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Can we add some doc here?

File: core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java [hunk 5]
```diff
@@ -208,12 +212,52 @@ public boolean supportsAliasedValues() {
...
+   * Unparses String concatenation operator for Mysql, Mysql uses Concat as string concatenation
+   * operator when PIPES_AS_CONCAT is not enabled. CONCAT concatenation characters are used by
+   * default.
+   *
+   * @param writer writer
+   * @param call call
+   */
+  private void unparseConcat(SqlWriter writer, SqlCall call) {
+    List<SqlNode> expandOperandList = new ArrayList<>();
+    expandConcatOperands(call, expandOperandList);
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java]: Why do we need to expand operands? `operand.unparse` can recursively handle nested case in nature.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 4 above)
Comment (wenhuitang) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Thanks a lot, I have already addressed.

File: core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java (hunk 5 above)
Comment (wenhuitang) [reply core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java]: Expand operands for CONCAT, so we can generate CONCAT(`fname`, '_', '_', `lname`) instead of CONCAT(CONCAT(CONCAT(`fname`, '_'), '_'), `lname`)

Comment (wenhuitang) [reply core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java]: > Why do we need to expand operands? `operand.unparse` can recursively handle nested case in nature.

Expand operands for CONCAT, so we can generate CONCAT(fname, '', '', lname) instead of CONCAT(CONCAT(CONCAT(fname, ''), ''), lname)

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 6]
```diff
@@ -286,6 +286,9 @@ private SqlLibraryOperators() {
               opBinding -> {
                 int precision = opBinding.collectOperandTypes().stream()
                     .mapToInt(RelDataType::getPrecision).sum();
+                // precision may be smaller than -1 when operands' precision are not be specified,
+                // which is not allowed.
+                precision = Math.max(precision, -1);
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: In `concat(x, y, z)`, if the precision of `x`, `y`, `z` are specified (i.e., precision != `PRECISION_NOT_SPECIFIED`), the return type is the sum of them. However, if one operand's precision is `PRECISION_NOT_SPECIFIED`, the return type should also be `PRECISION_NOT_SPECIFIED`. 
E.g., for `concat('aaa', cast('bb' as varchar), 'ccc')`, operand's precisions are: [3, -1, 3]. The return type's precision (3-1+3=5) is not resonable.

**Comment by DonnyZone** [comment]
Some comments:
(1)In Oracle, `concat` function accepts only two operands. In `SqlLibraryOperators`, We'd better seperate the `CONCAT_FUNCTION` into `CONCAT_FUNCTION` and `ORACLE_CONCAT_FUNCTION`. For  `ORACLE_CONCAT_FUNCTION`, we can use `SqlOperandCountRanges.of(2)` to enable the constraint.
(2)Implement the `CONCAT_FUNCTION` in SqlFunctions and register the implementation in `RexImpTable`.
(3)Update the unparse logic accordingly.

**Comment by wenhuitang** [comment]
> (1)In Oracle, `concat` function accepts only two operands. In `SqlLibraryOperators`, We'd better seperate the `CONCAT_FUNCTION` into `CONCAT_FUNCTION` and `ORACLE_CONCAT_FUNCTION`. For `ORACLE_CONCAT_FUNCTION`, we can use `SqlOperandCountRanges.of(2)` to enable the constraint.
How about seperate the `CONCAT_FUNCTION` into `CONCAT_FUNCTION`  which uses  SqlOperandCountRanges.from(3) and `CONCAT2` which uses `SqlOperandCountRanges.of(2)`? And the libraries for  `CONCAT_FUNCTION` are PG and MySql. The libraries for `CONCAT2` are Oracle, PG and MySQL .Since currently `CONCAT_FUNCTION`  which use SqlOperandCountRanges.from(2) contains the case of concat2 that may lead to confusion.

**Comment by DonnyZone** [comment]
There is no confusion. From user's perspective, when the library is specified, the function can be registered correctly. These two functions are placed in two distinct slots.

**Comment by wenhuitang** [comment]
> There is no confusion. From user's perspective, when the library is specified, the function can be registered correctly. These two functions are placed in two distinct slots.

According to you comments, this pull request has been updated, thanks a lot.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 7]
```diff
@@ -460,6 +460,14 @@ public static ByteString concat(ByteString s0, ByteString s1) {
     return s0.concat(s1);
   }
 
+  /** SQL {@code concat(string0, string2, string3, ...)} function. */
+  public static String concat(String... args) {
+    List<String> argList = Arrays.asList(args);
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: You can use `String.join("", args)`.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 8]
```diff
@@ -332,6 +334,8 @@
         NullPolicy.STRICT);
     defineMethod(CONCAT, BuiltInMethod.STRING_CONCAT.method,
         NullPolicy.STRICT);
+    defineMethod(CONCAT2, BuiltInMethod.STRING_CONCAT.method, NullPolicy.STRICT);
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: Maybe there is no need to register `CONCAT2` into `RexImpTable`. In runtime, Calcite can automatically find the concat function. You can have a try.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 9]
```diff
@@ -460,6 +460,14 @@ public static ByteString concat(ByteString s0, ByteString s1) {
     return s0.concat(s1);
   }
 
+  /** SQL {@code concat(string0, string2, string3, ...)} function. */
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: To align with `args`, "arg0, arg1..." is better than "string0, .." in doc.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 10]
```diff
@@ -655,6 +655,54 @@ public int size() {
...
+   * concat(cast('a' as varchar(65535)), cast('b' as varchar(2)), cast('c' as varchar(2)))
+   * returns varchar
+   */
+  public static final SqlReturnTypeInference MULTIVALENT_STRING_SUM_PRECISION =
+      opBinding -> {
+        boolean hasPrecisionNotSpecifiedOperand = false;
+        int typePrecision;
+        long amount = 0;
+        List<RelDataType> operandTypes = opBinding.collectOperandTypes();
+        for (RelDataType operandType: operandTypes) {
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: You can directly break the iteration when: 1) encountering `PRECISION_NOT_SPECIFIED`, 2) `amount` large than max precision.

**Comment by DonnyZone** [comment]
Moreover, (1) add several tests in `functions.iq`; (2) please revise the commit message & JIRA title & PR title as "Implement Concat function".

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 11]
```diff
@@ -655,6 +655,54 @@ public int size() {
         return ret;
       };
 
+
+  /**
+   * Type-inference strategy for String concatenation.
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: Improve the comment, you can refer to `DYADIC_STRING_SUM_PRECISION` for "||" operator.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 8 above)
Comment (wenhuitang) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: Still keep defineMethod for CONCAT2 for looking up method clearly.

**Comment by wenhuitang** [comment]
Comments has been addressed, and reference.md has been updated too.

**Comment by DonnyZone** [comment]
+1, LGTM

**Comment by DonnyZone** [comment]
Close and reopen to trigger the checks.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 12]
```diff
@@ -2169,6 +2173,31 @@ protected static Calendar getCalendarNotTooNear(int timeUnit) {
...
+          "VARCHAR(3) NOT NULL");
+      sqlTester.checkString("concat(cast('a' as varchar), cast('b' as varchar), "
+          + "cast('c' as varchar))", "abc", "VARCHAR NOT NULL");
+      sqlTester.checkNull("concat('a', 'b', cast(null as char(2)))");
+      sqlTester.checkNull("concat(cast(null as ANY), 'b', cast(null as char(2)))");
+    }
+    tester3.setFor(SqlLibraryOperators.CONCAT2);
+    tester3.checkString("concat(cast('fe' as char(2)), cast('df' as varchar(65535)))",
+        "fedf", "VARCHAR NOT NULL");
+    tester3.checkString("concat(cast('fe' as char(2)), cast('df' as varchar))",
```
Comment (chunweilei) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Please add a test case whose parameter is empty.

**Comment by DonnyZone** [comment]
In [CALCITE-3235](https://issues.apache.org/jira/browse/CALCITE-3235), `CONCAT` function with var args is added but not implemented (i.e., the query can only pass validation).
This PR focuses on:
(1) adding Oracle's `CONCAT` function, it accepts only two operands
(2) implementing `CONCAT` function in runtime
(3) fixing the return type inference issue

Personally, I think "Implement CONCAT function" is ok. 
@chunweilei @wenhuitang Do you have any suggestions to cover the topics above?

**Comment by danny0405** [comment]
> In [CALCITE-3235](https://issues.apache.org/jira/browse/CALCITE-3235), `CONCAT` function with var args is added but not implemented (i.e., the query can only pass validation).
> This PR focuses on:
> (1) adding Oracle's `CONCAT` function, it accepts only two operands
> (2) implementing `CONCAT` function in runtime
> (3) fixing the return type inference issue
> 
> Personally, I think "Implement CONCAT function" is ok.
> @chunweilei @wenhuitang Do you have any suggestions to cover the topics above?

"Supports CONCAT for variable arguments" seems fine to me.

**Comment by wenhuitang** [comment]
@DonnyZone @danny0405 @chunweilei Thanks a lot. I have addressed all the comments. And I'm ok with "Supports CONCAT for variable arguments", I will revise the relevant content.
